### PR TITLE
tikv: add the user's guide of stale read and safe ts

### PR DESCRIPTION
**Pull request by Oreoxmt**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

- Add the description of tikv-ctl subcommand `get-region-read-progress`
- Add a guide to diagnose issues related to stale read and safe-ts.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [ ] master (the latest development version)
- [ ] v7.4 (TiDB 7.4 versions)
- [x] v7.3 (TiDB 7.3 versions)
- [ ] v7.2 (TiDB 7.2 versions)
- [x] v7.1 (TiDB 7.1 versions)
- [ ] v7.0 (TiDB 7.0 versions)
- [x] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/14524, https://github.com/pingcap/docs/pull/15174
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: grafana-tidb-dashboard.md [hunk 1]
```diff
@@ -109,11 +109,22 @@ aliases: ['/docs-cn/dev/grafana-tidb-dashboard/','/docs-cn/dev/reference/key-mon
 
 ### KV Request
 
+下面的监控指标与发送给 TiKV 的请求相关。重试请求会被计数多次。
```
Comment (hfxsd) [inline grafana-tidb-dashboard.md:112]:
```suggestion
下面的监控指标与发送给 TiKV 的请求相关。重试请求会被多次计数。
```

File: grafana-tikv-dashboard.md [hunk 2]
```diff
@@ -404,6 +404,20 @@ aliases: ['/docs-cn/dev/grafana-tikv-dashboard/','/docs-cn/dev/reference/key-mon
 - Total pessimistic locks memory size：内存悲观锁占用内存的总大小
 - In-memory pessimistic locking result：将悲观锁仅保存到内存的结果，其中 full 表示因为超过内存限制而无法将悲观锁保存至内存的次数
 
+### Resolved-TS
+
+- Resolved-TS worker CPU：resolved-ts worker 线程的 CPU 使用率
+- Advance-TS worker CPU：advance-ts worker 线程的 CPU 使用率
+- Scan lock worker CPU：scan lock worker 线程的 CPU 使用率
+- Max gap of resolved-ts：在当前 TiKV 中，所有活跃 Region 的 resolved-ts 与当前时间的最大差值
```
Comment (hfxsd) [inline grafana-tikv-dashboard.md:412]: 截图上的名称是 Max Resolved TS gap
```suggestion
- Max Resolved TS gap：在当前 TiKV 中，所有活跃 Region 的 resolved-ts 与当前时间的最大差值
```

Comment (Oreoxmt) [reply grafana-tikv-dashboard.md:412]: @ekexium PTAL and update the English document if necessary.

File: troubleshoot-stale-read.md
Comment (Oreoxmt) [inline troubleshoot-stale-read.md:1]: Preview: https://scaffold-git-preview-pingcapdocs-cn15365-oreoxmt.vercel.app/zh/tidb/dev/troubleshoot-stale-read

File: troubleshoot-stale-read.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    tracked index: 92,
+    number of locks: 0,
+    number of transactions: 0,
+    stopped: false,
+```
+
+上面的输出结果可以帮助你判断：
+
+- 锁是否阻塞了 resolved-ts。
+- apply index 是否太小而无法更新 safe-ts。
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:100]:
```suggestion
- `applied index` 是否太小而无法更新 safe-ts。
```

File: troubleshoot-stale-read.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+- 当存在 follower peer 时，leader 是否发送了足够更新的 resolved-ts。
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:101]:
```suggestion
- 当存在 follower peer 时，leader 是否发送了更新的 resolved-ts。
```

File: troubleshoot-stale-read.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+- 在 `tikv-ctl` 命令中指定 `--log` 选项，并在 TiKV 日志中通过 start_ts 查找相应的锁。
+- 在 TiDB 和 TiKV 日志中搜索 start_ts，以识别事务的问题。
+
+    如果一个查询花费超过 60 秒，就会打印一个带有 SQL 语句的 `expensive_query` 日志。你可以使用 start_ts 值匹配日志。下面是一个示例：
+
+    ```log
+    [2023/07/17 19:32:09.403 +08:00] [WARN] [expensivequery.go:145] [expensive_query] [cost_time=60.025022732s] [cop_time=0.00346666s] [process_time=8.358409508s] [wait_time=0.013582596s] [request_count=278] [total_keys=9943616] [process_keys=9943360] [num_cop_tasks=278] [process_avg_time=0.030066221s] [process_p90_time=0.045296042s] [process_max_time=0.052828934s] [process_max_addr=192.168.31.244:20160] [wait_avg_time=0.000048858s] [wait_p90_time=0.00006057s] [wait_max_time=0.00040991s] [wait_max_addr=192.168.31.244:20160] [stats=t:442916666913587201] [conn=2826881778407440457] [user=root] [database=test] [table_ids="[100]"] [**txn_start_ts**=442916790435840001] [mem_max="2514229289 Bytes (2.34 GB)"] [sql="update t set b = b + 1"]
+    ```
+
+- Use the [`CLUSTER_TIDB_TRX`](/information-schema/information-schema-tidb-trx.md#cluster_tidb_trx) table to find active transactions if you cannot get enough information about the locks from logs.
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:142]: 这句话漏译了？

Comment (hfxsd) [inline troubleshoot-stale-read.md:142]:
```suggestion
```

File: troubleshoot-stale-read.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+- 如果你无法从日志中获取关于锁的足够信息，可以使用 [`CLUSTER_TIDB_TRX`](/information-schema/information-schema-tidb-trx.md#cluster_tidb_trx) 表查找活跃的事务。
+- 执行 [`SHOW PROCESSLIST`](/sql-statements/sql-statement-show-processlist.md) 查看当前连接到同一个 TiDB 服务器的会话及其在当前语句上花费的时间。但是它不会显示 start_ts。
+
+如果锁是由于正在进行的大事务而存在的，考虑修改你的应用程序逻辑，因为这些锁会阻碍 resolve-ts 的进度。
+
+如果锁不属于任何正在进行的事务，可能是由于协调器 (TiDB) 在预写锁之后崩溃。在这种情况下，TiDB 会自动解决锁。除非问题持续存在，否则不需要采取任何措施。
+
+### 处理长事务
+
+长时间保持活跃的事务可能会阻塞 resolved-ts 的进度，即使它们最终很快就提交了。这是因为这些长期存在的事务的 start-ts 用于计算 resolved-ts。
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:152]:
```suggestion
长时间保持活跃的事务，即使最终提交了，也可能会阻塞 resolved-ts 的进度。这是因为这些长期存在的事务的 start-ts 用于计算 resolved-ts。
```

File: grafana-tikv-dashboard.md (hunk 2 above)
Comment (ekexium) [reply grafana-tikv-dashboard.md:412]: In older version it's "Max Resolved TS gap". It's been renamed. The pic is captured from an older version.

File: troubleshoot-stale-read.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+- apply index 是否太小而无法更新 safe-ts。
+- 当存在 follower peer 时，leader 是否发送了足够更新的 resolved-ts。
+
+### 使用日志诊断
+
+TiKV 每 10 秒检查以下监控项：
+
+- resolved-ts 最小的 Region leader
+- safe-ts 最小的 Region follower
+- resolved-ts 最小的 Region follower
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:109]:
```suggestion
- resolved-ts 最小的 Region leader
- resolved-ts 最小的 Region follower
- safe-ts 最小的 Region follower
```

File: troubleshoot-stale-read.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+Resolver:
+    exist: true,
+    resolved_ts: 442918444145049601,
+    tracked index: 2477,
+    number of locks: 480000,
+    number of transactions: 1,
+    stopped: false,
+```
+
+值得注意的是，`applied_index` 等于 resolver 中的 `tracked index`。因此，resolver 可能是这个问题的根源。你还可以看到，有 1 个事务在这个 Region 中留下了 480000 个锁，这可能是问题的原因。
```
Comment (hfxsd) [inline troubleshoot-stale-read.md:208]:
```suggestion
值得注意的是，`applied_index` 等于 resolver 中的 `tracked index`，均为 `2477`。因此，resolver 可能是这个问题的根源。你还可以看到，有 1 个事务在这个 Region 中留下了 480000 个锁，这可能是问题的原因。
```

File: troubleshoot-stale-read.md (hunk 3 above)
Comment (Oreoxmt) [reply troubleshoot-stale-read.md:100]: @ekexium PTAL

Comment (ekexium) [reply troubleshoot-stale-read.md:100]: 这个应该就是 apply index

File: troubleshoot-stale-read.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+safe-ts 是一个由 Region 中的每个 peer 维护的时间戳，它保证所有时间戳小于等于 safe-ts 的事务已经被 peer apply，从而实现本地 Stale Read。
+
+### 什么是 resolved-ts？
+
+resolved-ts 是一个时间戳，它保证所有时间戳小于该值的事务已经被 leader apply。与 safe-ts 不同，resolved-ts 只由 Region leader 维护。Follower 可能有一个比 leader 更小的 apply index，因此 resolved-ts 不能直接被当作 safe-ts。
+
+### safe-ts 的维护
+
+`RegionReadProgress` 模块维护 safe-ts。Region leader 维护 resolved-ts，并定期通过 CheckLeader RPC 将 resolved-ts、最小的 apply index（用于验证 resolved-ts）和 Region 本身发送给所有副本的 `RegionReadProgerss` 模块。
```
Comment (ekexium) [inline troubleshoot-stale-read.md:28]:
```suggestion
`RegionReadProgress` 模块维护 safe-ts。Region leader 维护 resolved-ts，并定期通过 CheckLeader RPC 将 resolved-ts、最小的（使 resolved-ts 生效的）apply index和 Region 本身发送给所有副本的 `RegionReadProgerss` 模块。
```

File: troubleshoot-stale-read.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+[2023/08/29 16:48:18.118 +08:00] [INFO] [endpoint.rs:526] ["the max gap of follower safe-ts is large"] [oldest_candidate=None] [latest_candidate=None] [applied_index=3276] [duration_to_last_consume_leader=11460ms] [resolved_ts=443888117117353985] [safe_ts=443888117117353985] [gap=111503ms] [region_id=273]
+
+[2023/08/29 16:48:18.118 +08:00] [INFO] [endpoint.rs:547] ["the max gap of follower resolved-ts is large; it's the same region that has the min safe-ts"]
+```
+
+## 诊断建议
+
+### 处理慢事务提交
+
+提交时间长的事务通常是大事务。这个慢事务的 prewrite 阶段会留下一些锁，但是在 commit 阶段清理锁之前需要很长时间。为了解决这个问题，你可以尝试识别锁所属的事务，并找出它们存在的原因，例如使用日志。
```
Comment (ekexium) [inline troubleshoot-stale-read.md:129]:
```suggestion
提交时间长的事务通常是大事务。这个慢事务的 prewrite 阶段会留下一些锁，但是在 commit 阶段清理掉锁之前需要很长时间。为了解决这个问题，你可以尝试识别锁所属的事务，并找出它们存在的原因，例如使用日志。
```

File: troubleshoot-stale-read.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+值得注意的是，`applied_index` 等于 resolver 中的 `tracked index`，均为 `2477`。因此，resolver 可能是这个问题的根源。你还可以看到，有 1 个事务在这个 Region 中留下了 480000 个锁，这可能是问题的原因。
+
+为了获取确切的事务和一些锁的 keys，你可以检查 TiKV 日志并搜索 `locks with`。输出结果如下：
+
+```log
+[2023/07/17 21:16:44.257 +08:00] [INFO] [resolver.rs:213] ["locks with the minimum start_ts in resolver"] [keys="[74800000000000006A5F7280000000000405F6, ... , 74800000000000006A5F72800000000000EFF6, 74800000000000006A5F7280000000000721D9, 74800000000000006A5F72800000000002F691]"] [start_ts=442918429687808001] [region_id=3121]
+```
+
+从 TiKV 日志中，你可以获取事务的 start_ts，即 `442918429687808001`。为了获取关于语句和事务的更多信息，你可以在 TiDB 日志中搜索 `start_ts`。输出结果如下：
```
Comment (ekexium) [inline troubleshoot-stale-read.md:215]:
```suggestion
从 TiKV 日志中，你可以获取事务的 start_ts，即 `442918429687808001`。为了获取关于语句和事务的更多信息，你可以在 TiDB 日志中搜索这个时间戳。找到结果如下：
```

**Review by ekexium — COMMENTED** [review]
英文的我也对应改了一下

**Review by ekexium — APPROVED** [review]

File: grafana-tikv-dashboard.md (hunk 2 above)
Comment (hfxsd) [reply grafana-tikv-dashboard.md:412]: Should we update this pic? [media/stale-read/example-ts-gap.png](https://github.com/pingcap/docs-cn/pull/15365/files#diff-834ac55aacb61d2fb3b62ed7a7fe430a3ee1ed7161f6c6183253a79f62f2ebfb)

Comment (ekexium) [reply grafana-tikv-dashboard.md:412]: Not necessary. We can comment the rename here

**Review by hfxsd — APPROVED** [review]

**Pull request merged**